### ISSUE TITLE
ProteinBERTs embeddings are inconsistent between batch sizes and not independent of other sequences in the batch

### ISSUE THREAD

**Issue by felbecker** [nadavbra/protein_bert#59]
ProteinBERT computes slightly varying embeddings for the same input sequence, when this sequence is processed as part of a batch and when it is processed alone. The padding is the same in both cases. There seems to be some (perhaps only numeric) dependence between elements in a batch, which looks unsual and wrong to me.

This code reproduces the issue:

```
from proteinbert import load_pretrained_model
from proteinbert.conv_and_global_attention_model import get_model_with_hidden_layers_as_outputs
import numpy as np


seqs = ["ACGT", "CTG"]
                    
                    
pretrained_model_generator, input_encoder = load_pretrained_model()
model = get_model_with_hidden_layers_as_outputs(pretrained_model_generator.create_model(32))

#run the model on the batch
encoded_x = input_encoder.encode_X(seqs, 32)
local_representations, _ = model(encoded_x)
                    
#run the model on just the first sequence
encoded_x_0 = input_encoder.encode_X([seqs[0]], 32)
local_representations_0, _ = model(encoded_x_0)

np.testing.assert_almost_equal(local_representations[0, :4].numpy(), local_representations_0[0, :4].numpy())
```

The output is:

Arrays are not almost equal to 7 decimals

```
Mismatched elements: 5214 / 6248 (83.5%)
Max absolute difference: 0.00156593
Max relative difference: 0.55014306
 x: array([[ 8.8389345e-02,  9.5654741e-02, -1.6907825e-01, ...,
         1.0000000e+00,  1.8658639e-11,  9.3445485e-18],
       [ 1.5440202e-01,  5.0813537e-03,  2.2378229e-02, ...,...
 y: array([[ 8.8389367e-02,  9.5654681e-02, -1.6907825e-01, ...,
         1.0000000e+00,  1.8545951e-11,  9.2982521e-18],
       [ 1.5440196e-01,  5.0813323e-03,  2.2378221e-02, ...,...
```

Can you tell what's going on here?

**Comment by nadavbra** [comment]
Thank you @felbecker for reporting this!
If the embeddings/results from ProteinBERT are batch-dependent, that would indeed be a major issue. It seems that the differences you show indeed seem very small, suggesting that it may indeed be a numerical issue (or quirks of the hardware).

Does it matter if you do the calculation on a GPU or a CPU? Do you ever get more significant differences? If you run the exact same batch twice, do you get the exact same results?

**Comment by felbecker** [comment]
Thank you for your quick reply.

The embeddings are consistent for multiple runs of the same batch. 

After opening this issue, however, I realized that other Transformer based language models like ESM2 or ProtT5 also have similarly small absolute differences of the embeddings. So it seems to be a normal thing. 

FYI: Finetuning on my downstream task works fine though. I only discovered this when I was writing a unit test for my language model wrapper class. I figured I could test if the embeddings are correctly computed if I would compare them to the embeddings I get if I just feed a single sequence without any extra stuff. The unit tests pass for proteinBERT (and the others) if I just decrease the decimals to 3.

I will try to reproduce on CPU or other GPUs.
